### PR TITLE
esp32: add espidf_heap_info to esp32 module

### PR DESCRIPTION
**Pull request by tve**
This PR adds a `espidf_heap_info()` method to the esp32 module which provides info about the ESP-IDF heaps. It's useful to get a bit of a picture of what's going on when stuff starts failing because esp-idf can't allocate memory anymore. See docs for details and a sample output.
I have not tested with esp-idf v3.

### CODE REVIEW

File: ports/esp32/modesp32.c [hunk 1]
```diff
@@ -145,6 +148,32 @@ STATIC mp_obj_t esp32_hall_sensor(void) {
...
+STATIC mp_obj_t esp32_espidf_heap_info(void) {
+    #if 0
+    heap_caps_print_heap_info(MALLOC_CAP_8BIT);
+    return mp_obj_new_int(0);
+    #else
+    multi_heap_info_t info;
+    heap_t *heap;
+    mp_obj_t heap_list = mp_obj_new_list(0, 0);
+    SLIST_FOREACH(heap, &registered_heaps, next) {
+        if (heap_caps_match(heap, MALLOC_CAP_8BIT)) {
```
Comment (dpgeorge) [inline ports/esp32/modesp32.c:160]: Would it be useful to list all memory types?  eg would be good to know available IRAM for loading native code in to.  In that case probably need to add an extra entry to the return tuple, at the start, being the memory type.

File: ports/esp32/modesp32.c [hunk 2]
```diff
@@ -153,6 +182,7 @@ STATIC const mp_rom_map_elem_t esp32_module_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR_wake_on_ext1), MP_ROM_PTR(&esp32_wake_on_ext1_obj) },
     { MP_ROM_QSTR(MP_QSTR_raw_temperature), MP_ROM_PTR(&esp32_raw_temperature_obj) },
     { MP_ROM_QSTR(MP_QSTR_hall_sensor), MP_ROM_PTR(&esp32_hall_sensor_obj) },
+    { MP_ROM_QSTR(MP_QSTR_espidf_heap_info), MP_ROM_PTR(&esp32_espidf_heap_info_obj) },
```
Comment (dpgeorge) [inline ports/esp32/modesp32.c:185]: Since the function is in the `esp32` module, maybe just `heap_info` is enough, or `idf_heap_info`?

**Review by dpgeorge — COMMENTED** [review]
I think this is a good addition.

**Comment by dpgeorge** [comment]
> I have not tested with esp-idf v3.

That's fine.  As long as it builds it should be ok.

File: ports/esp32/modesp32.c (hunk 2 above)
Comment (tve) [reply ports/esp32/modesp32.c:185]: `idf_heap_info` would be fine with me, I just wanted to make sure someone new doesn't think this is the MP heap...

Comment (dpgeorge) [reply ports/esp32/modesp32.c:185]: sounds good

**Comment by tve** [comment]
Reworked and added a test. Works in esp-idf v3 and v4. Ready as far as I'm concerned.

I renamed to idf_heap_info and I added a parameter, which is the esp-idf MEM_CAP_XXX bitfield. There are two constants in the esp32 module that select what MP uses for data and for code but advanced users can pass in any bitfield they desire.

CI fails 'cause the arm minimal build increased by 4 bytes even though none of the code I changed is included, I believe CI needs some love ;-)

**Comment by tve** [comment]
Force-pushed to remove HW CI commit that erroneously made it in. Double-checked, good to go from my point of view.

File: docs/library/esp32.rst [hunk 3]
```diff
@@ -88,6 +112,10 @@ Constants
 
     Used in `Partition.find` to specify the partition type.
 
+.. data:: MEM_DATA
+          MEM_EXEC
```
Comment (dpgeorge) [inline docs/library/esp32.rst:116]: Maybe call these constants `HEAP_DATA` and `HEAP_EXEC` to indicate they are related to the `idf_heap_info()` function, the key word being "heap"?

File: tests/esp32/esp32_idf_heap_info.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+try:
+    import esp32
+except ImportException:
+    print("SKIP")
+    raise SystemExit
+
+# try getting heap regions, looks like a couple of small regions are always taken?
+heap = esp32.idf_heap_info(esp32.MEM_DATA)
+# print("heap:", heap)
+if heap[1] == (7288, 0, 0, 0):
```
Comment (dpgeorge) [inline tests/esp32/esp32_idf_heap_info.py:10]: Will these specific numbers remain stable across different esp32 builds, or IDF versions?  It won't work to update this test each time the IDF version changes, or IDF configuration changes (SDK configuration which may change initial memory layout).

With these kinds of things the tests usually just check the length of the tuple and types of entries.  Here, it should be possible to test that the first 3 entries are always `>=` the next one.

Also, tests usually just print out their results (they don't check values in the test), and the .exp file contains the correct values which are tested against.

**Comment by dpgeorge** [comment]
I see this was force-pushed.  Is it ready for re-review/merge?

**Comment by tve** [comment]
Yes, it's ready! Sorry for not being more explicit.

File: tests/esp32/esp32_idf_heap_info.py.exp [hunk 5]
```diff
@@ -0,0 +1,5 @@
+HEAP_DATA >3: True
```
Comment (dpgeorge) [inline tests/esp32/esp32_idf_heap_info.py.exp:1]: I think this should be `HEAP_DATA >5: True`

File: tests/esp32/esp32_idf_heap_info.py.exp [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+HEAP_DATA [True, True, True, True]
+HEAP_EXEC >3: True
```
Comment (dpgeorge) [inline tests/esp32/esp32_idf_heap_info.py.exp:3]: `HEAP_EXEC >2: True`

**Review by dpgeorge — COMMENTED** [review]
Apart from the two minor changes to the test exp, it looks good.

**Comment by tve** [comment]
My bad, thanks for catching the test failure. I fixed the .exp as you suggested and ran it on an esp32 with MP configured to exclude bluetooth support. Guess what... it comes up with 4 data heap regions! So I set both tests to `>2` and am crossing fingers.

**Comment by dpgeorge** [comment]
Squashed and merged in a177831c46a7e217fed6562510e12183377053f7 with some minor changes: docs literal code formatting fixed, some cross-docs links added, fixed `ImportException` to `ImportError` in test.

**Comment by tve** [comment]
Thanks much for the last fixes! I'll make an effort not to leave crap in the next PR...

**Pull request closed without merging**